docs/design/features/readytorun-elf-support-roadmap.md

```diff
@@ -0,0 +1,166 @@
+  and section tables and emitting the section data in the proper place.
+
+* **Diagnostics / SOS**: switching over to native ELF will have fundamental effect on
+  diagnostics. While the overall long-term effect is assumed to be positive,
+  it will take time to bridge the user experience gap. It is likely that for some
+  temporary period of time we'll need to keep support for both PE files and ELF
+  files on Linux until all diagnostic functionality for ELF has been put on par
+  with pre-existing diagnostic support for PE-based executables.
+
+* **Visual Studio, VS Code**: switching over the executable format is certain to
+  have non-trivial effect on both VS and VS Code. We need to plan ahead with the
+  relevant teams to make sure we cover all the relevant project template, build
+  system, debugger, profiler and other aspects.
+
+* **R2RDump**: The PE header reader that is today using the managed assembly
+  System.Reflection.PortableExecutable needs to be generalized to recognize and
+  parse the ELF headers and locate the RTR_HEADER export symbol - once we're able
+  to identify the R2R header in the file, we immediately have access to all the
+  sections and data. Naturally we'll also need to adjust the current logic for
+  translating RVA's to file offsets.
+
+* **ILSpy**: will hopefully continually ingest R2RDump changes so it should
+  automatically pick up support for ELF-embedded MSIL/ECMA-335 metadata.
+
+# Proposed sequencing
+
+The natural sequencing and initial T-shirt costing implied by the various
+inter-dependencies and by the natural requirement that each step needs to be
+verifiable / diagnosable is as follows:
+
+* Import (rewrite) the relevant ELF headers to managed code, probably
+  somewhere under ``src/coreclr/src/tools/Common/Internal/Runtime``.
+  This is junior level work estimated at 2 man-weeks including some rudimentary
+  regression testing - e.g. writing a simple test that will dump a subset of
+  the headers for a hard-coded ELF file used as a resource by the test.
+  
+  An interesting observation is that most ELF header structures have
+  parallel 32-bit and 64-bit variants. It might be useful to logically unify them
+  by internally using longs and just have virtual de-serializers that would have
+  something like ReadNativeInt representing 32-bit or 64-bit integer based on
+  the bitness.
+  
+  We probably don't need to support ELF to the fullest, e.g. endianness doesn't
+  matter as our JIT and runtime doesn't support any high-endian CPU's (ARM has
+  technically switchable endianness but we only support the little variant).
+
+* Create a simplified counterpart of System.Reflection.PortableExecutable to
+  support ELF, again somewhere under ``src/coreclr/src/tools/Common/Internal/Runtime``.
+  At some point we may consider porting it to libraries but initially it's better
+  to keep that separate.
+  
+  We should also double-check whether mono doesn't already have something in this
```

Comment:
Haven't looked at this in detail, but this came up on my Twitter feed recently: https://github.com/xoofx/LibObjectFile. It also comes with some level of DWARF support.